railties/lib/rails/generators/rails/app/templates/config/puma.rb.tt

```diff
@@ -2,38 +2,53 @@
 # are invoked here are part of Puma's configuration DSL. For more information
 # about methods provided by the DSL, see https://puma.io/puma/Puma/DSL.html.
 
-# Puma can serve each request in a thread from an internal thread pool.
-# The `threads` method setting takes two numbers: a minimum and maximum.
-# Any libraries that use thread pools should be configured to match
-# the maximum value specified for Puma. Default is set to 5 threads for minimum
-# and maximum; this matches the default thread size of Active Record.
-max_threads_count = ENV.fetch("RAILS_MAX_THREADS") { 5 }
-min_threads_count = ENV.fetch("RAILS_MIN_THREADS") { max_threads_count }
-threads min_threads_count, max_threads_count
-
-# Specifies that the worker count should equal the number of processors in production.
-if ENV["RAILS_ENV"] == "production"
+rails_env = ENV.fetch("RAILS_ENV", "development")
+
+# Puma starts a configurable number of processes (workers) and each process
+# serves each request in a thread from an internal thread pool.
+#
+# As a rule of thumb, increasing the number of threads will increase how much
+# traffic a given process can handle (throughput), but due to Ruby's
+# Global VM Lock (GVL) it has diminishing returns, and will degrade the
+# response time (latency) of the application.
+#
+# The ideal number of threads per worker depend both how much time the
+# application spend waiting for IO operations, and how much you wish to
+# to prioritize throughput over latency.
+#
+# The default is set to 3 threads as it's deemed a decent compromise between
+# throughput and latency for the average Rails application.
+#
+# Any libraries that use a connection pool or another resource pool should
+# be configured to provide at least as many connections as the number of
+# threads. This includes Active Record's `pool` parameter in `database.yml`.
+threads_count = ENV.fetch("RAILS_MAX_THREADS") { 3 }
+threads threads_count, threads_count
```

Comment:
In the future we'll be able to only pass a single value: https://github.com/puma/puma/pull/3309 (❤️ @dentarg)

Reply:
Assuming that'll be in puma > 6.4.2, could you add a conditional here? Something like

```suggestion
<%= PUMA_VERSION > "6.4.2" ? "threads threads_count" : "threads threads_count, threads_count" %>
```

![image](https://github.com/rails/rails/assets/509837/c8ac113a-a0a2-43b6-a549-5714ae80fb9f)

Reply:
Files are generated before the gems are installed to the system. We can't version check on puma.